src/ConnectionAutomaton.cpp

```diff
@@ -172,22 +186,64 @@ void ConnectionAutomaton::onConnectionFrame(Payload payload) {
     case FrameType::KEEPALIVE: {
       Frame_KEEPALIVE frame;
       if (frame.deserializeFrom(std::move(payload))) {
-        assert(frame.header_.flags_ & FrameFlags_KEEPALIVE_RESPOND);
-        frame.header_.flags_ &= ~(FrameFlags_KEEPALIVE_RESPOND);
-        connectionOutput_.onNext(frame.serializeOut());
+        if (!client_) {
+          if (frame.header_.flags_ & FrameFlags_KEEPALIVE_RESPOND) {
+            frame.header_.flags_ &= ~(FrameFlags_KEEPALIVE_RESPOND);
+            connectionOutput_.onNext(frame.serializeOut());
+          } else {
+            Frame_ERROR errorFrame(
+                0,
+                ErrorCode::INVALID,
+                folly::IOBuf::copyBuffer("keepalive without flag"));
+            connectionOutput_.onNext(errorFrame.serializeOut());
+            // TODO(yschimke) should this be onTerminal
+            cancel();
+          }
+        }
+        // TODO(yschimke) client *should* check the respond flag
       } else {
-        // TODO(yschimke): handle connection-level error
-        assert(false);
+        Frame_ERROR errorFrame(
+            0,
+            ErrorCode::INVALID,
+            folly::IOBuf::copyBuffer("unexpected frame"));
+        connectionOutput_.onNext(errorFrame.serializeOut());
+        // TODO(yschimke) should this be onTerminal
+        cancel();
       }
     }
       return;
-    case FrameType::SETUP:
-      // TODO handle lease logic
-      LOG(INFO) << "ignoring setup frame";
+    case FrameType::SETUP: {
+      Frame_SETUP frame;
+      if (frame.deserializeFrom(std::move(payload))) {
+        if (frame.header_.flags_ & FrameFlags_LEASE) {
+          // We don't have the correct lease and wait logic above yet
+          Frame_ERROR errorFrame(
+              0,
+              ErrorCode::UNSUPPORTED_SETUP,
+              folly::IOBuf::copyBuffer("leases not supported"));
+          connectionOutput_.onNext(errorFrame.serializeOut());
+          // TODO(yschimke) should this be onTerminal
+          cancel();
+        }
+      } else {
+        // TODO(yschimke) make this conditional if we have versioning problems
+        Frame_ERROR errorFrame(
+            0,
+            ErrorCode::INVALID_SETUP,
+            folly::IOBuf::copyBuffer("bad setup frame"));
+        connectionOutput_.onNext(errorFrame.serializeOut());
+        // TODO(yschimke) should this be onTerminal
+        cancel();
+      }
+    }
       return;
     default:
-      // TODO(yschimke): check ignore flag and fail
-      assert(false);
+      // TODO(yschimke) make this conditional if we have versioning problems
+      Frame_ERROR errorFrame(
```

Comment:
this is a second time when I see this error frame with the same message.
Should we create static methods for such instances?
Eq.
Frame_ERROR::unexpectedFrame() which would return a new instance of Frame_ERROR?
similarly
Frame_ERROR::badSetupFrame(), 
etc...

For improved readability..